src/gobupload/apply/main.py

```diff
@@ -18,38 +18,36 @@
 ANALYZE_THRESHOLD = 0.3
 
 
-def apply_events(storage, last_events, start_after, stats):
+def apply_events(storage: GOBStorageHandler, last_events: set[str], start_after: int, stats: UpdateStatistics):
     """Apply any unhandled events to the database
 
     :param storage: GOB (events + entities)
+    :param last_events: all entities with events applied
     :param start_after: the is of the last event that has been applied to the storage
     :param stats: update statitics for this action
     :return:
     """
-    with ActiveGarbageCollection("Apply events"), storage.get_session() as session:
-        logger.info("Apply events")
+    PROCESS_PER = 10_000
 
-        PROCESS_PER = 10000
+    with (
+        ActiveGarbageCollection("Apply events"),
+        ProgressTicker("Apply events", PROCESS_PER) as progress,
+        storage.get_session()
+    ):
+        logger.info("Apply events")
         add_event_tids = set()
-        with ProgressTicker("Apply events", PROCESS_PER) as progress:
-            unhandled_events = storage.get_events_starting_after(start_after, PROCESS_PER)
-            while unhandled_events:
-                with EventApplicator(storage) as event_applicator:
-                    for event in unhandled_events:
-                        progress.tick()
-
-                        gob_event, count, applied_events = event_applicator.apply(
-                            event, last_events, add_event_tids)
-                        action = gob_event.action
-                        stats.add_applied(action, count)
-                        start_after = event.eventid
 
-                        # Remove event from session, to avoid trying to update event db object
-                        session.expunge(event)
+        while unhandled_events := list(storage.get_events_starting_after(start_after, PROCESS_PER)):
```

Comment:
Why list() and not next? By using list() we still load everything into memory, thereby undoing the streaming in the get_events_starting_after method.

Seems like we're using two different strategies for limiting memory usage here as well. On one hand we stream the results, on the other hand we have manual limiting (paging) of the result set.